sycl/include/sycl/ext/oneapi/experimental/enqueue_functions.hpp

```diff
@@ -0,0 +1,327 @@
+}
+
+template <typename KernelName = sycl::detail::auto_name, int Dimensions,
+          typename KernelType, typename... ReductionsT>
+void nd_launch(queue Q, nd_range<Dimensions> Range, const KernelType &KernelObj,
+               ReductionsT &&...Reductions) {
+  submit(Q, [&](handler &CGH) {
+    nd_launch(CGH, Range, KernelObj, std::forward<ReductionsT>(Reductions)...);
+  });
+}
+
+template <typename KernelName = sycl::detail::auto_name, int Dimensions,
+          typename Properties, typename KernelType, typename... ReductionsT>
+void nd_launch(handler &CGH,
+               launch_config<nd_range<Dimensions>, Properties> Config,
+               const KernelType &KernelObj, ReductionsT &&...Reductions) {
+
+  ext::oneapi::experimental::detail::LaunchConfigAccess<nd_range<Dimensions>,
+                                                        Properties>
+      ConfigAccess(Config);
+  CGH.parallel_for<KernelName>(ConfigAccess.getRange(),
+                               std::forward<ReductionsT>(Reductions)...,
+                               KernelObj);
+}
+
+template <typename KernelName = sycl::detail::auto_name, int Dimensions,
+          typename Properties, typename KernelType, typename... ReductionsT>
+void nd_launch(queue Q, launch_config<nd_range<Dimensions>, Properties> Config,
+               const KernelType &KernelObj, ReductionsT &&...Reductions) {
+  submit(Q, [&](handler &CGH) {
+    nd_launch(CGH, Config, KernelObj, std::forward<ReductionsT>(Reductions)...);
+  });
+}
+
+template <int Dimensions, typename... ArgsT>
+void nd_launch(handler &CGH, nd_range<Dimensions> Range,
+               const kernel &KernelObj, ArgsT &&...Args) {
+  CGH.set_args<ArgsT...>(std::forward<ArgsT>(Args)...);
+  CGH.parallel_for(Range, KernelObj);
+}
+
+template <int Dimensions, typename... ArgsT>
+void nd_launch(queue Q, nd_range<Dimensions> Range, const kernel &KernelObj,
+               ArgsT &&...Args) {
+  submit(Q, [&](handler &CGH) {
+    nd_launch(CGH, Range, KernelObj, std::forward<ArgsT>(Args)...);
+  });
+}
+
+template <int Dimensions, typename Properties, typename... ArgsT>
+void nd_launch(handler &CGH,
+               launch_config<nd_range<Dimensions>, Properties> Config,
+               const kernel &KernelObj, ArgsT &&...Args) {
+  ext::oneapi::experimental::detail::LaunchConfigAccess<nd_range<Dimensions>,
+                                                        Properties>
+      ConfigAccess(Config);
+  CGH.set_args<ArgsT...>(std::forward<ArgsT>(Args)...);
+  CGH.parallel_for(ConfigAccess.getRange(), KernelObj);
+}
+
+template <int Dimensions, typename Properties, typename... ArgsT>
+void nd_launch(queue Q, launch_config<nd_range<Dimensions>, Properties> Config,
+               const kernel &KernelObj, ArgsT &&...Args) {
+  submit(Q, [&](handler &CGH) {
+    nd_launch(CGH, Config, KernelObj, std::forward<ArgsT>(Args)...);
+  });
+}
+
+inline void memcpy(handler &CGH, void *Dest, const void *Src, size_t NumBytes) {
```

Comment:
Just out of curiosity, since the one of the objective of this extension is to simplify SYCL abstractions, why don't we have functions like `memcpy_async` and `memcpy_sync`, as opposed to just `memcpy`? As a novice to SYCL, I personally find that explicit `async` or `sync` suffixes to memory operations makes it less error-prone for the users.

Reply:
I also believe that all existing SYCL commands are non-blocking.  I think teaching developers to use everything asynchronously can actually be beneficial (to performance) because it means they're more likely to write code like:

```c++
sycl::memcpy(q, ...);
sycl::memcpy(q, ...);
sycl::barrier(q);
sycl::nd_launch(q, ...).;
q.wait(); // host-device synchronization
```

...which allows the execution of the two `memcpy` calls to overlap (on hardware with that capability), and defers host-device synchronization until its actually necessary.

If we started introducing synchronous `memcpy` calls, I think novice users would be more likely to gravitate to those because they seem "simpler", leading them to write this instead:

```c++
sycl::memcpy_sync(q, ...); // host-device synchronization
sycl::memcpy_sync(q, ...); // host-device synchronization
sycl::nd_launch(...);
q.wait(); // host-device synchronization
```

Reply:
I see. That makes sense.